### PR TITLE
Creating phone mnemonics with recursion 

### PR DESCRIPTION
**Pull request by orlovska**
Hi! Could you please review my solution for this problem.

### CODE REVIEW

File: python_code/recursion_phone_mnemonics.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
+# From input number(7 to 9 characters) return all possible phone mnemonics 
+# Digits corresponding characters: ('0', '1', 'ABC', 'DEF', 'GHI', 'JKL', 'MNO', 'PQRS', 'TUV', 'WXYZ')
+
+
+def chars_for_digit(number):
+    List = ['0', '1', 'ABC', 'DEF', 'GHI', 'JKL', 'MNO', 'PQRS', 'TUV', 'WXYZ']
```
Comment (cluePrints) [inline python_code/recursion_phone_mnemonics.py]: vars names are generally lower case

File: tests/test_recursion_phone_mnemonics.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        self.assertEqual(phone_mnemonics("1"), ['1'])
+        
+    def test_long_phone_number(self):
+        self.assertEqual(phone_mnemonics("23"), ['AD', 'AE', 'AF', 'BD', 'BE', 'BF', 'CD', 'CE', 'CF'])
+        self.assertEqual(phone_mnemonics("234"), ['ADG', 'ADH', 'ADI', 'AEG', 'AEH', 'AEI', 'AFG', 'AFH', 'AFI', 'BDG', 'BDH', 'BDI', 'BEG', 'BEH', 'BEI', 'BFG', 'BFH', 'BFI', 'CDG', 'CDH', 'CDI', 'CEG', 'CEH', 'CEI', 'CFG', 'CFH', 'CFI'])
+        self.assertEqual(phone_mnemonics("2345"), ['ADGJ', 'ADGK', 'ADGL', 'ADHJ', 'ADHK', 'ADHL', 'ADIJ', 'ADIK', 'ADIL', 'AEGJ', 'AEGK', 'AEGL', 'AEHJ', 'AEHK', 'AEHL', 'AEIJ', 'AEIK', 'AEIL', 'AFGJ', 'AFGK', 'AFGL', 'AFHJ', 'AFHK', 'AFHL', 'AFIJ', 'AFIK', 'AFIL', 'BDGJ', 'BDGK', 'BDGL', 'BDHJ', 'BDHK', 'BDHL', 'BDIJ', 'BDIK', 'BDIL', 'BEGJ', 'BEGK', 'BEGL', 'BEHJ', 'BEHK', 'BEHL', 'BEIJ', 'BEIK', 'BEIL', 'BFGJ', 'BFGK', 'BFGL', 'BFHJ', 'BFHK', 'BFHL', 'BFIJ', 'BFIK', 'BFIL', 'CDGJ', 'CDGK', 'CDGL', 'CDHJ', 'CDHK', 'CDHL', 'CDIJ', 'CDIK', 'CDIL', 'CEGJ', 'CEGK', 'CEGL', 'CEHJ', 'CEHK', 'CEHL', 'CEIJ', 'CEIK', 'CEIL', 'CFGJ', 'CFGK', 'CFGL', 'CFHJ', 'CFHK', 'CFHL', 'CFIJ', 'CFIK', 'CFIL'])
```
Comment (cluePrints) [inline tests/test_recursion_phone_mnemonics.py]: must have been pretty exhausting to populate 😅 

hint: for permutation tests & large inputs one of the strategies is to just check the total count + a few cases (like 1st/last)

File: python_code/recursion_phone_mnemonics.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+                mnemonic = mnemonic[:-1]
+                                                    # After the end of for loop - CHANGITG SUPERIOR MNEMONIC LETTER
+        mnemonic = mnemonic[:-1]
+                                                    # COMPLETE ONE RECURSIVE CALL  
+        if check_for_more_values(phone_number, result) is False:
+            return result
+        else:
+            return mnemonic
+    
+                                                    # RECURSION BASE CASE
```
Comment (cluePrints) [inline python_code/recursion_phone_mnemonics.py]: comments are usually not  caps (caps is screaming), also in this case if it's related to `if phone_number == '''` - usually the `#` is aligned to be just on top of `i`

File: python_code/recursion_phone_mnemonics.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+            nonlocal mnemonic
+            mnemonic += digit
+                                                    # END OF MNEMONIC - when phone_number decreased to one digit whitch had been added to mnemonic          
+            if len(phone_number) > 1:
+                phone_mnemonics_helper(phone_number[1:])
+            else:
+                result.append(mnemonic)
+                                                    # Atfer adding to result - CHANGING LAST MNEMONIC LETTER
+                mnemonic = mnemonic[:-1]
+                                                    # After the end of for loop - CHANGITG SUPERIOR MNEMONIC LETTER
```
Comment (cluePrints) [inline python_code/recursion_phone_mnemonics.py]: nit: changiNg

File: python_code/recursion_phone_mnemonics.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+def phone_mnemonics(phone_number):
+    if type(phone_number) not in [str]:
+        raise TypeError("Required argument in quotes: '', '234'")
+    
+    result = []
+    mnemonic = ''
+    
+    def phone_mnemonics_helper(phone_number):
+        for digit in chars_for_digit(phone_number[0]):
+            nonlocal mnemonic
```
Comment (cluePrints) [inline python_code/recursion_phone_mnemonics.py]: I think if mnemonic will be a list and not a string, you'll be able to pass it as the function param and get rid of nonlocal thing

File: python_code/recursion_phone_mnemonics.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+            if len(phone_number) > 1:
+                phone_mnemonics_helper(phone_number[1:])
+            else:
+                result.append(mnemonic)
+                                                    # Atfer adding to result - CHANGING LAST MNEMONIC LETTER
+                mnemonic = mnemonic[:-1]
+                                                    # After the end of for loop - CHANGITG SUPERIOR MNEMONIC LETTER
+        mnemonic = mnemonic[:-1]
+                                                    # COMPLETE ONE RECURSIVE CALL  
+        if check_for_more_values(phone_number, result) is False:
```
Comment (cluePrints) [inline python_code/recursion_phone_mnemonics.py]: bool constants are usually not used in if conditions, e.g. here:

```
if not check_for_more_values(phone_number, result):
    ...
```

File: python_code/recursion_phone_mnemonics.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    
+    result = []
+    mnemonic = ''
+    
+    def phone_mnemonics_helper(phone_number):
+        for digit in chars_for_digit(phone_number[0]):
+            nonlocal mnemonic
+            mnemonic += digit
+                                                    # END OF MNEMONIC - when phone_number decreased to one digit whitch had been added to mnemonic          
+            if len(phone_number) > 1:
```
Comment (cluePrints) [inline python_code/recursion_phone_mnemonics.py]: I get a sense that `len(phone_number) > 1` is basically checking if mnemonic is not yet complete.

Might be a bit cleaner this if would be focused on results:

```
if len(mnemonic) == expected_length:
    # looks like a complete mnemonic - add to the results list
    result.append(mnemonic)
    ...
else:
    # we need moar letters
    phone_mnemonics_helper(phone_number[1:])
```

File: python_code/recursion_phone_mnemonics.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    return List[int(number)] 
+
+
+def check_for_more_values(phone_number, result):
+    try:
+        last_char_of_the_supirior_node = chars_for_digit(phone_number[0])
+        if last_char_of_the_supirior_node[-1] in result[-1]:
+            return False
+    except:
+        return True
```
Comment (cluePrints) [inline python_code/recursion_phone_mnemonics.py]: curious when does this happen? using try/except for regular control flow is generally frowned upon

File: python_code/recursion_phone_mnemonics.py [hunk 9]
```diff
@@ -0,0 +1,49 @@
+# From input number(7 to 9 characters) return all possible phone mnemonics 
+# Digits corresponding characters: ('0', '1', 'ABC', 'DEF', 'GHI', 'JKL', 'MNO', 'PQRS', 'TUV', 'WXYZ')
+
+
+def chars_for_digit(number):
+    List = ['0', '1', 'ABC', 'DEF', 'GHI', 'JKL', 'MNO', 'PQRS', 'TUV', 'WXYZ']
+    return List[int(number)] 
+
+
+def check_for_more_values(phone_number, result):
```
Comment (cluePrints) [inline python_code/recursion_phone_mnemonics.py]: boolean funcs often start with `is_` or `has_` (might be a Java thing - be mindful of me not being a pythonista), e.g. `has_more_values`

File: python_code/recursion_phone_mnemonics.py [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+    try:
+        last_char_of_the_supirior_node = chars_for_digit(phone_number[0])
+        if last_char_of_the_supirior_node[-1] in result[-1]:
```
Comment (cluePrints) [inline python_code/recursion_phone_mnemonics.py]: so we're getting last of the results and checking that the last of the chars for the last number is in there, might result in a bug for inputs using the same number multiple times, e.g: `222`

File: python_code/recursion_phone_mnemonics.py [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+    
+    def phone_mnemonics_helper(phone_number):
+        for digit in chars_for_digit(phone_number[0]):
+            nonlocal mnemonic
+            mnemonic += digit
+                                                    # END OF MNEMONIC - when phone_number decreased to one digit whitch had been added to mnemonic          
+            if len(phone_number) > 1:
+                phone_mnemonics_helper(phone_number[1:])
+            else:
+                result.append(mnemonic)
```
Comment (cluePrints) [inline python_code/recursion_phone_mnemonics.py]: I also wonder if extracting this out of the cycle and changing the conditions around might make the whole thing a bit easier to follow - `check_for_more_values` won't be needed

File: python_code/recursion_phone_mnemonics.py (hunk 5 above)
Comment (orlovska) [reply python_code/recursion_phone_mnemonics.py]: I've tried it. Got an "UnboundLocalError: local variable 'mnemonic' referenced before assignment". 
However, the result list still working in the inner function..

**Pull request merged**